### DISCUSSION TITLE
EventSubscriptionExample seems doesn't work?

### DISCUSSION THREAD

**Discussion by deamn** [eclipse-milo/milo#1008] [General]
I try to execute EventSubscriptionExample with a Kepware KEPServerEX OPCUA server, but it seems that Event Subscription desn't work? Nothing appends despite setValueConsumer return value changed. I also tested with/without EventFilter.
Do you have some tips?

**Comment by kevinherron** [comment]
KEPServerEX doesn't publish any events as far as I'm aware.

**Reply by deamn** [reply]
Maybe with KEPServerEX I need the Alarms and Events plugin (https://www.kepware.com/en-us/products/kepserverex/advanced-plug-ins/alarms-events/).
I also tested EventSubscriptionExample with vNode OPCUA Server (https://vnodeautomation.com) and Siemens SimaticNET OPCUA server but nothing is happening, the connection and the variable subscription seems ok. Could you explain the expected behavior of EventSubscriptionExample?

**Reply by kevinherron** [reply]
The expected behavior is that _if_ a server publishes events it will receive them and log some of the basic event properties. You can run it against the Milo example server to see this in action. It should also work against the Prosys demo server.

Its primary purpose is to show an example of how to create an EventFilter and create a MonitoredItem that will receive events.

**Reply by deamn** [reply]
Thanks for the explaination. In between, I have also tested with Milo server 0.6.6 (ExampleServer), nothing is happening. I tested also with Identifiers.Server_ServerStatus_CurrentTime (instead of Identifiers.Server), an event should be generated each seconds, right? Monday I will test with Prosys demo.
I need to make something special to generate event on server?

**Reply by kevinherron** [reply]
No, the example server generates them automatically, but it sounds like you don't really understand the difference between MonitoredItems for events and MonitoredItems for other attribute changes.

`Server_ServerStatus_CurrentTime` is a Variable so it's not even a valid target for an event subscription. Events are monitored by creating a MonitoredItem that targets the EventNotifier attribute of an Object Node where that attribute has the SubscribeToEvents bit set to 1.

The Server object acts as the default event notifier for most servers, _if_ they support events, but in my experience most servers so far do not (that's starting to change though).

When I run the EventSubscriptionExample as it is, which automatically starts the example server to connect to, I see this output:
```
08:26:52.253 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - Event Received from NodeId{ns=0, id=2253}
08:26:52.253 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[0]: ByteString{bytes=[0, 1, 2, 3]}
08:26:52.253 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[1]: NodeId{ns=0, id=2041}
08:26:52.253 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[2]: 2
08:26:52.253 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[3]: DateTime{utcTime=132999532122060000, javaDate=Fri Jun 17 08:26:52 PDT 2022}
08:26:52.253 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[4]: LocalizedText{text=event message!, locale=en}
08:26:54.271 [milo-shared-thread-pool-1] INFO  o.e.m.e.c.EventSubscriptionExample - Event Received from NodeId{ns=0, id=2253}
08:26:54.271 [milo-shared-thread-pool-1] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[0]: ByteString{bytes=[0, 1, 2, 3]}
08:26:54.271 [milo-shared-thread-pool-1] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[1]: NodeId{ns=0, id=2041}
08:26:54.271 [milo-shared-thread-pool-1] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[2]: 2
08:26:54.271 [milo-shared-thread-pool-1] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[3]: DateTime{utcTime=132999532142210000, javaDate=Fri Jun 17 08:26:54 PDT 2022}
08:26:54.271 [milo-shared-thread-pool-1] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[4]: LocalizedText{text=event message!, locale=en}
08:26:56.290 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - Event Received from NodeId{ns=0, id=2253}
08:26:56.290 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[0]: ByteString{bytes=[0, 1, 2, 3]}
08:26:56.290 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[1]: NodeId{ns=0, id=2041}
08:26:56.290 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[2]: 2
08:26:56.290 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[3]: DateTime{utcTime=132999532162270000, javaDate=Fri Jun 17 08:26:56 PDT 2022}
08:26:56.290 [milo-shared-thread-pool-2] INFO  o.e.m.e.c.EventSubscriptionExample - 	variant[4]: LocalizedText{text=event message!, locale=en}
```

If you wanted to see the CurrentTime changing every second you're looking in the wrong place. Try one of these:
- [SubscriptionExample](https://github.com/eclipse/milo/blob/master/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/SubscriptionExample.java)
- [ManagedSubscriptionDataExample](https://github.com/eclipse/milo/blob/master/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/ManagedSubscriptionDataExample.java)
